http/src/v0/refs.rs

```diff
@@ -138,32 +138,253 @@ async fn refs_paths<T: IpfsTypes>(
+        }
+    }
+}
+
+impl std::error::Error for WalkError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use WalkFailed::*;
+        match &self.reason {
+            Loading(_) => None, // TODO: anyhow
+            Parsing(e) => Some(e),
+            DagPb(e) => Some(e),
+            IpldWalking(e) => Some(e),
+        }
+    }
+}
+
+impl From<Error> for WalkFailed {
+    fn from(e: Error) -> Self {
+        WalkFailed::Loading(e)
+    }
+}
+
+impl From<libipld::error::BlockError> for WalkFailed {
+    fn from(e: libipld::error::BlockError) -> Self {
+        WalkFailed::Parsing(e)
+    }
+}
+
+impl From<ipfs::unixfs::ll::ResolveError> for WalkFailed {
+    fn from(e: ipfs::unixfs::ll::ResolveError) -> Self {
+        WalkFailed::DagPb(e)
+    }
+}
+
+impl From<path::WalkFailed> for WalkFailed {
+    fn from(e: path::WalkFailed) -> Self {
+        WalkFailed::IpldWalking(e)
+    }
+}
+
+impl From<(WalkFailed, Cid)> for WalkError {
+    fn from((reason, last_cid): (WalkFailed, Cid)) -> Self {
+        WalkError { last_cid, reason }
+    }
+}
+
+/// The IpfsPath walk can end in with the target block loaded or parsed and optionally projected as
+/// an Ipld.
+#[derive(Debug)]
+pub enum Loaded {
+    /// The raw block from `ipfs.get_block`
+    Raw(Box<[u8]>),
+    /// Possibly projected IPLD value.
+    Ipld(Ipld),
+}
+
 /// Walks the `path` while loading the links.
 ///
-/// # Panics
-///
-/// If there are dag-pb nodes and the libipld has changed it's dag-pb tree structure.
+/// Returns the Cid where we ended up, and an optional Ipld structure if one was projected, and the
+/// path inside the last document we walked.
 pub async fn walk_path<T: IpfsTypes>(
     ipfs: &Ipfs<T>,
     mut path: IpfsPath,
-) -> Result<(Cid, Ipld), Error> {
+) -> Result<(Cid, Loaded, Vec<String>), WalkError> {
+    use ipfs::unixfs::ll::MaybeResolved;
+
     let mut current = path.take_root().unwrap();
 
-    loop {
-        let Block { data, .. } = ipfs.get_block(&current).await?;
-        let ipld = decode_ipld(&current, &data)?;
+    // cache for any datastructure used in repeated hamt lookups
+    let mut cache = None;
+
+    // the path_inside_last applies only in the IPLD projection case and it's main consumer is the
+    // `/dag/resolve` API where the response is the returned cid and the "remaining path".
+    let mut path_inside_last = Vec::new();
+
+    // important: on the `/refs` path we need to fetch the first block to fail deterministic so we
+    // need to load it either way here; if the response gets processed to the stream phase, it'll
+    // always fire up a response and the test 'should print nothing for non-existent hashes' fails.
+    // Not sure how correct that is, but that is the test.
+    'outer: loop {
+        let Block { data, .. } = match ipfs.get_block(&current).await {
+            Ok(block) => block,
+            Err(e) => return Err(WalkError::from((WalkFailed::from(e), current))),
```

Comment:
wouldn't `Err(WalkError::from((e.into(), current)))` work here (and possibly in any other similar places)?

Reply:
It might work, I try to avoid calling `.into()` as in some historic rust version (or my own errors) I've felt that it might be an easier compilation error.

With the [above proposed](https://github.com/rs-ipfs/rust-ipfs/pull/184/files#r438757057) this would become `return Err(e.at(current));`

But I'll happily not expand this unless it really makes this more readable :)